### DISCUSSION TITLE
Is it a good idea to make a `Scaffold` widget globally available (e.g. through a provider)?

### DISCUSSION THREAD

**Discussion by fabis94** [rrousselGit/riverpod#1331] [Q&A] [answered]
In the `graphql_flutter` package there's a way to catch network level errors globally (through `ErrorLink`), and when these happen I'd like to throw a Snackbar message on the screen. But since this `ErrorLink` is defined outside of any widgets, it doesn't have access to any kind of `BuildContext`.

So I'm wondering if it would be a bad idea to store the main `Scaffold` instance into some kind of global/singleton state (Provider?), and then access it from the `ErrorLink` callback to be able to show error messages on the screen.

If not, how do I solve this issue in the correct way? I don't want to duplicate network error checking code everywhere inside widgets that actually execute GraphQL operations.

**Comment by scopendo** [comment, marked as answer]
I would not make a `Scaffold` available via Riverpod, however, it's perfectly OK to make a scaffold messenger key available via Riverpod and then derive the `ScaffoldMessenger` from that.

```dart
/// A provider whose value is the global scaffold messenger key passed to
/// [MaterialApp].
///
final scaffoldMessengerKeyPod = Provider((ref) {
  return GlobalKey<ScaffoldMessengerState>();
});

/// Returns the scaffold messenger associated with [scaffoldMessengerKeyPod].
///
final scaffoldMessengerPod = Provider((ref) {
  return ref.watch(scaffoldMessengerKeyPod).currentState!;
});

...

final scaffoldMessenger = ref.read(scaffoldMessengerPod);

scaffoldMessenger.showSnackBar(
  SnackBar(
    content: Text('Some message'),
  ),
);
```

**Reply by fabis94** [reply]
Didn't know it was possible to do it this way, nice

**Reply by IsaiasCuvula** [reply]
yeah This is a good idea I guess, or you can use a global key as shown here: https://api.flutter.dev/flutter/material/ScaffoldMessenger/of.html#material.ScaffoldMessenger.of.2
